flang/lib/Lower/Runtime.cpp

```diff
@@ -348,18 +348,31 @@ void Fortran::lower::genSystemClock(fir::FirOpBuilder &builder,
                                     mlir::Location loc, mlir::Value count,
                                     mlir::Value rate, mlir::Value max) {
   auto makeCall = [&](mlir::FuncOp func, mlir::Value arg) {
+    mlir::Type type = arg.getType();
+    fir::IfOp ifOp{};
+    if (type.dyn_cast<fir::PointerType>() || type.dyn_cast<fir::HeapType>()) {
```

Comment:
I think you also need to handle the case where `arg` is a dummy optional. Using `fir::valueHasFirAttribute(arg,  fir::getOptionalAttrName())` should allow you to check that.

```
subroutine foo(i, j)
  integer, optional :: i, j
  call system_clock(count=i, count_rate=j)
end subroutine
```

Reply:
Done; thanks.